### PR TITLE
[master] Add audio soloing feature

### PR DESCRIPTION
**Pull request by Atlante45**
https://highfidelity.manuscript.com/f/cases/19553/Add-audio-sniper-solo-feature-to-efficiently-ask-the-mixer-for-one-or-more-specific-sources

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14763/HighFidelity-Beta-PR14352-37c69eb.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14763/HighFidelity-Beta-PR14352-37c69eb.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14763/HighFidelity-Beta-Interface-PR14352-37c69eb.dmg

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14763/HighFidelity-Beta-Interface-PR14352-37c69eb.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14791/HighFidelity-Beta-PR14352-4e802b0.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14791/HighFidelity-Beta-Interface-PR14352-4e802b0.dmg

**Comment by Atlante45** [comment]
build this please

**Comment by MiladNazeri** [comment]
Hi @Atlante45 !  Thank you so much for the work on this.  This is great.  

One note is that our JSDOC's for the scripting API is formatted to be Javascript-centric, so the areas where type is QVector<QUuid>, could that be changed to match getPalData which take a similar array of UUIDs?  

Reference:

https://github.com/highfidelity/hifi/blob/fac9188dfb6b83c561f75038c0b83039c0e9203b/interface/src/avatar/AvatarManager.h#L187

https://highfidelity.atlassian.net/wiki/spaces/HOME/pages/331612165/Hifi+JSDoc+Guide

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14798/HighFidelity-Beta-PR14352-4e802b0.apk

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14802/HighFidelity-Beta-PR14352-0ca0a6f.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14798/HighFidelity-Beta-PR14352-4e802b0.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14798/HighFidelity-Beta-Interface-PR14352-4e802b0.dmg

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14798/HighFidelity-Beta-Interface-PR14352-4e802b0.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14802/HighFidelity-Beta-PR14352-0ca0a6f.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14802/HighFidelity-Beta-Interface-PR14352-0ca0a6f.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14802/HighFidelity-Beta-PR14352-0ca0a6f.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14802/HighFidelity-Beta-Interface-PR14352-0ca0a6f.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14811/HighFidelity-Beta-PR14352-b8cb433.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14811/HighFidelity-Beta-PR14352-b8cb433.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14811/HighFidelity-Beta-Interface-PR14352-b8cb433.dmg

**Comment by Atlante45** [comment]
## Test Plan:

Pre-requisite: You will need 3 people for this test.
Person `A` is running the test.
Person `B` will be the solo target.
Person `C` will act as a control

1) Run the Sandbox and gather all 3 people in the domain.
2) `B` should change their display name to "SoloTarget" (the case matters)
3) `A` should make sure that everyone can hear each other.
4) `A` should now start the following script:
https://gist.githubusercontent.com/Atlante45/4e2f433e580eb84334a7d7e92da75199/raw/29ddf8b80dda33d1325c148085d03519dbf166d5/SoloTest.js
5) From now on `A` should only be able to hear `B`
6) `B` should not be attenuated. (Move away from `B` and they should stay at the max volume.
7) Kill the script and you should now hear both `B` and `C` again.

Note: A couple more things to test while the solo is on in step 6):
- If `B` ignores(/mutes) `A`, `A` should not hear `B` even if they are soloing `B`, except if `A` is an admin.
- The bubble should not affect the soloing.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14865/HighFidelity-Beta-PR14352-65a83cb.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14865/HighFidelity-Beta-PR14352-65a83cb.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14865/HighFidelity-Beta-Interface-PR14352-65a83cb.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14865/HighFidelity-Beta-PR14352-65a83cb.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14865/HighFidelity-Beta-Interface-PR14352-65a83cb.exe

**Comment by Atlante45** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14897/HighFidelity-Beta-PR14352-ef54a63.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14897/HighFidelity-Beta-PR14352-ef54a63.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14897/HighFidelity-Beta-Interface-PR14352-ef54a63.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14897/HighFidelity-Beta-PR14352-ef54a63.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14897/HighFidelity-Beta-Interface-PR14352-ef54a63.exe

File: libraries/audio/src/AudioSolo.cpp [hunk 1]
```diff
@@ -0,0 +1,88 @@
+//
+//  AudioSolo.cpp
+//
```
Comment (birarda) [inline libraries/audio/src/AudioSolo.cpp]: Put the path to this file in the header

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14938/HighFidelity-Beta-PR14352-f337579.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14938/HighFidelity-Beta-PR14352-f337579.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14938/HighFidelity-Beta-Interface-PR14352-f337579.dmg

**Comment by antoninahifi** [comment]
currently testing

**Comment by qlipdahlin** [comment]
Currently Testing

**Comment by qlipdahlin** [comment]
@conklin94122 PASS

Tested with 3 people and on sandbox. Testers could hear each other before script was run and Tester B set name to SoloTarget. Script was started and could only hear Tester B, while B and C could hear all Testers fine.

Tested attenuation with Tester C following Tester A as they move away from Tester B whom continued to speak. Audio for Tester A continued at full volume while tester C heard less and eventually could not hear Tester B.

Tested mute, with Tester A unable to hear Tester B when they are muted, save when they are an admin and People app is open.

Tested bubble with no bubble, Tester A bubble only, Tester B bubble only, Tester A and B with bubble. Tester A could hear Tester B with bubble on, Tester B could not hear Tester A with bubble on, and Tester C could hear everyone. - Atlante45 confirmed this is fine.

PR Passes.

**Pull request merged**